Manage behaviour of trying to create more than one XMPP client.

lucyhe opened the pull request:
- Remove all client listeners on logout
- If we call connect when a client already exists, log out the old client first
Tested manually with Chrome and Firefox and I believe that it will fix https://github.com/uProxy/uproxy/issues/1370
At least, this prevents a few module crashes I've been able to reproduce.

ryscheng commented on src/core/socialprovider.js:
```diff
@@ -122,6 +122,14 @@ XMPPSocialProvider.prototype.initializeState = function() {
  * @param {Function} continuation Callback upon connection
  */
 XMPPSocialProvider.prototype.connect = function(continuation) {
+  if (this.client) {
```
Do we have a sense of when this is triggered? My intuition is that if connect() is called when this.client exists, that's representative of a bug.

lucyhe replied:
This can definitely happen if the user clicks one of the login buttons more than once (how I ended up reproducing the mentioned bug). I think it can also happen with reconnect (how I am guessing Izzie reproduced the bug). 

In the first case, it's because the social provider's login method is called more than once. We could stop connect() from being called twice by keeping track of pending or successful login attempts in the UI, core.ts and/or social.ts; however, that would prevent chrome_oauth.ts from being able to open a new auth tab each time the button is clicked (since the tab is launched after the social provider's login method is called).

In the reconnect case, it's possible that the old client is not cleaned up before the new connect attempt starts.

I have a hunch that changing the last few lines of google-auth/facebook-auth to:
```
if (!this.client) {
  this.initializeState();
  this.connect(continuation);
}
```
or changing the beginning of XMPPSocialProvider's onCredentials to:
```
if (msg.cmd && msg.cmd === 'auth') {
  this.credentials = msg.message;
  this.connect(continuation);
}
```
could prevent connect being called multiple times. I felt a little better editing the connect method, but I can explore these changes as well.

ryscheng commented on src/core/socialprovider.js:
```diff
@@ -167,7 +175,6 @@ XMPPSocialProvider.prototype.connect = function(continuation) {
 
     if (this.client) {
       this.client.end();
-      delete this.client;
```
This block is confusing me now. Error events can happen at any time, not just as a result of login. So we should decide to do 1 of 3 things:
a. Do a full cleanup if an error event happens and log out.
b. Generally ignore the error
c. Figure out which errors can be returned and appropriatedly do either a or b depending on the severity.

But just ending the session without updating the appropriate local state seems like the wrong thing to do?

dborkan replied:
My understanding of this code is that calling ```this.client.end()``` will result in in an "end" event being fire which results in logout being called (lines 195-208).  The reason why ```delete this.client``` was removed is because that "end" callback could be invoked asynchronously, i.e. first we would call end(), then we would delete this.client, then the "end" callback would be invoked but because this.client wasn't set nothing would happen and the user would remain logged in but with this.client==null, leading to bugs.

Possibly we could just change the code to call logout() here rather than this.client.end()

And yes ideally we should do more to parse the errors and ignore the ones that aren't fatal or otherwise recover gracefully.  But I think until we know more it's best to log the user out in case of errors - otherwise they risk being logged in but nothing working.  Plus we have and will be improving automatic reconnect in the uProxy code so this will be less of a UX issue

ryscheng replied:
Awesome explanation! Thanks :)
I think calling logout() makes sense then to make sure we are cleaning up properly (plus its a bit more readable) 
Once we test that, I think this PR is good

lucyhe commented:
Just fixed the tests! It looks unclear from the diff what happened, but I added a beforeEach for all the tests that call .connect(). ptal, thank you!!

ryscheng commented on spec/socialprovider.spec.js:
```diff
@@ -326,4 +209,133 @@ describe("Tests for message batching in Social provider", function() {
         });
   });
+
+  describe("Tests for connection", function() {
```
Nested describes? Can we move this out or flatten somehow?

lucyhe commented:
Yeah of course! I felt this just put the missing piece in one place, but I can remove the second describe and set the client to null in each test that calls connect. Or I can take out line 59 (xmppSocialProvider.client = xmppClient;) and only initialize the client in the tests that do not call connect? Preferences?

ryscheng commented:
I think the latter makes sense. Thanks!

lucyhe commented:
done! thanks for suggestions

ryscheng commented:
:+1:
